File: 백준/Silver/2839. 설탕 배달/설탕 배달.py
n = int(input())

dp = [123456789] * (n + 1)

for i in range(3, n + 1):
    if i % 5 == 0:
        dp[i] = i // 5
    if i % 3 == 0:
        dp[i] = i // 3
    dp[i] = min(dp[i - 3] + 1, dp[i - 5] + 1, dp[i])

print(dp[n] if dp[n] < 10000 else -1)

File: 백준/Silver/2839. 설탕 배달/설탕 배달.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws Exception {
		BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(bufferedReader.readLine());

		int n = Integer.parseInt(st.nextToken());
		int[] arr = new int[n + 1];
		Arrays.fill(arr, 5000);

		for (int i = 3; i < n + 1; i++) {
			if (i % 3 == 0) {
				arr[i] = i / 3;
			}
			if (i % 5 == 0) {
				arr[i] = i / 5;
			}
			if (i > 4)
				arr[i] = Math.min(arr[i - 5] + 1, arr[i]);
			arr[i] = Math.min(arr[i], arr[i - 3] + 1);
		}

		if (arr[n] == 5000)
			System.out.println(-1);
		else
			System.out.println(arr[n]);

	}
}